### PR TITLE
[tune] TensorBoard HParams

### LINKED ISSUE #4528: [tune] Integration with TensorBoard HParams Dashboard
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux
- **Ray installed from (source or binary)**: binary
- **Ray version**: 0.6.2
- **Python version**: 2.7
- **Exact command to reproduce**: NA

### Describe the feature request
Would it be possible to see the `ray.tune`'s results in Tensorboard's HParams https://www.tensorflow.org/tensorboard/r2/hyperparameter_tuning_with_hparams dashboard?

### PR DESCRIPTION
**Pull request by vakker**
## Why are these changes needed?

Hyperparams logging in Tensorboard seems to be a great way to look through experiments.
By default TB shows all the logged `ray/tune/...` metrics, but they can be disabled in the UI, so it's not too disturbing if someone is not interested in all.

On the other hand, the table only shows the last value for the specified metrics, which might be a problem when the model is overtrained.
Maybe another metric could be added that tracks the lowest validation loss, but that should go to the custom implementation of the `Trainable` I think.

## Related issue number
Closes #4528 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.

### CODE REVIEW

**Comment by vakker** [comment]
@richardliaw pls let me know if this looks good or needs any further changes.

**Review by richardliaw — COMMENTED** [review]
Thanks a lot for this PR! I really appreciate it; and I think there's a better way to do this.

I've recently exposed a `trial.evaluated_params` field in `trial.py:Trial`. Can we use that instead to update the hparams used in TB? I haven't tested whether it is the right data type. 

So, to make this concrete, in Logger:

```python
# Expose `trial` to the Logger, optionally because you can use this logger outside Tune.
class Logger()
    def __init__(self, config, logdir, trial=None):
        self.config = config
        self.logdir = logdir
        self.trial = trial
        self._init()


# Send `trial` to all loggers used in Tune
class UnifiedLogger(Logger):
    def __init__(self, config, logdir, trial=None, loggers=None, sync_function=None):
       ...
    
    def _init(self):
        self._loggers = []
        for cls in self._logger_cls_list:
            try:
                self._loggers.append(cls(self.config, self.logdir, trial=self.trial))
               ...

# Keep most of your changes, use `trial` object fields.
class TF2Logger(Logger):
    ...
    # I think `evaluated_params` may not work here; you may need to check
    ... hp.hparams(self.trial.evaluated_params, trial_id=self.trial.trial_id)

```

To pass in the Trial object to the logger, you can do this in trial.py: 
```python

    def init_logger(self):
        ...
            self.result_logger = UnifiedLogger(
                self.config,
                self.logdir,
                trial=self,  # Pass the trial object here
                loggers=self.loggers,
                sync_function=self.sync_to_driver_fn)
```

Let me know if you have any questions!

**Comment by vakker** [comment]
Thanks for the suggestion, that does work. I changed the code, pls see the new version.

However, there's a problem with this approach, that the current code uses a string to store the params, e.g.: `lr=0.0001,momentum=0`, which I currently parse into `{'lr': '0.0001', 'momentum': '0'}`. But this will treat each parameter as a string, so on TB the plots and the hp filtering is not going to work.

Also, I noticed in `trial.py@138`:
```python
self.evaluated_params = evaluated_params or []
```
which suggests that it should be a list, not just a plain string. Maybe this is a bug?

In any case, what would you suggest to keep the right type for the parameters?

**Comment by richardliaw** [comment]
Ah, right - this looks like a bug. `evaluated_params` should be a dict. The purpose of that field was for TB HParams anyways, but I just didn't get to it previously.

File: doc/source/tune-usage.rst [hunk 1]
```diff
@@ -547,6 +547,23 @@ You can pass in your own logging mechanisms to output logs in custom formats as
...
+In order to use the hparam logging, add ``hparams`` and ``exp_id`` (optional) to the dict returned by the ``_train()`` method of a ``Trainable``, e.g.:
+
+.. code-block:: python
+    
+   def _train(self):
+        # training code
+        training_stats = {'training_loss': training_loss}
+        training_stats.update({'hparams': hparams})
+        training_stats.update({'exp_id'}: self._experiment_id)
+        return training_stats
```
Comment (richardliaw) [inline doc/source/tune-usage.rst]: We can probably get rid of this now that it's automatic.

**Review by richardliaw — COMMENTED** [review]
OK this looks good to me.

@vakker Thanks a bunch for starting this effort! I appreciate it. Could you try it out yourself and see if you can find any issues?

Also, can you touch up the docs to showcase this? Now that users don't need to do anything manually, I think it makes sense to append a hparams image in the Quick Start (under TensorBoard). Let me know what you think.

File: doc/source/tune-usage.rst (hunk 1 above)
Comment (vakker) [reply doc/source/tune-usage.rst]: Yep, I'll clean up the doc with some additional info and images as you mentioned below.

**Comment by vakker** [comment]
Thanks, it looks good to me.
I added a few tests as well to catch issues with `evaluated_params`.

Let me know if something is missing.

**Review by richardliaw — APPROVED** [review]
This looks good to me! I added a figure and another test. Will merge after tests pass

**Comment by richardliaw** [comment]
jenkins retest this please

**Comment by Nithanaroy** [comment]
Awesome!
What parts of this would work for pre TF 2.0?

**Comment by richardliaw** [comment]
Unfortunately right now this is isolated for TF2.0.

**Comment by vakker** [comment]
@richardliaw is there something else I need to do on this PR to make it pass the tests?

**Comment by richardliaw** [comment]
Sorry for the late response - just merged it!
